safekeeper/spec/ProposerAcceptorStatic.tla

```diff
@@ -0,0 +1,448 @@
+                    \* term histories with it to always have common point in
+                    \* them. Lsn is 1 because TLA+ sequences are indexed from 1
+                    \* (we don't want to truncate WAL out of range).
+                    termHistory |-> << [term |-> 0, lsn |-> 1] >>,
+                    wal |-> << >>
+                 ]]
+  /\ committed = {}
+  /\ elected_history = EmptyF
+
+
+\********************************************************************************
+\* Actions
+\********************************************************************************
+
+\* Proposer loses all state.
+\* For simplicity (and to reduct state space), we assume it immediately gets
+\* current state from quorum q of acceptors determining the term he will request
+\* to vote for.
+RestartProposer(p, q) ==
+  /\ Quorum(q)
+  /\ LET
+       new_term == Maximum({acc_state[a].term : a \in q}) + 1
+     IN
+       /\ prop_state' = [prop_state EXCEPT ![p].state = "campaign",
+                                           ![p].term = new_term,
+                                           ![p].votes = EmptyF,
+                                           ![p].termHistory = << >>,
+                                           ![p].wal = << >>,
+                                           ![p].nextSendLsn = EmptyF]
+       /\ UNCHANGED <<acc_state, committed, elected_history>>
+
+\* Term history of acceptor a's WAL: the one saved truncated to contain only <=
+\* local FlushLsn entries.
+AcceptorTermHistory(a) ==
+    SelectSeq(acc_state[a].termHistory, LAMBDA th_entry: th_entry.lsn <= FlushLsn(a))
+
+\* Acceptor a immediately votes for proposer p.
+Vote(p, a) ==
+ /\ prop_state[p].state = "campaign"
+ /\ acc_state[a].term < prop_state[p].term \* main voting condition
+ /\ acc_state' = [acc_state EXCEPT ![a].term = prop_state[p].term]
+ /\ LET
+      vote == [termHistory |-> AcceptorTermHistory(a), flushLsn |-> FlushLsn(a)]
+    IN
+      prop_state' = [prop_state EXCEPT ![p].votes = (a :> vote) @@ prop_state[p].votes]
+ /\ UNCHANGED <<committed, elected_history>>
+
+
+\* Get lastLogTerm from term history th.
+LastLogTerm(th) == th[Len(th)].term
+
+\* Proposer p gets elected.
+BecomeLeader(p) ==
+  /\ prop_state[p].state = "campaign"
+  /\ Quorum(DOMAIN prop_state[p].votes)
+  /\ LET
+       \* Find acceptor with the highest <last_log_term, lsn> vote.
+       max_vote_acc == CHOOSE a \in DOMAIN prop_state[p].votes:
+                         LET v == prop_state[p].votes[a]
+                         IN \A v2 \in Range(prop_state[p].votes):
+                              /\ LastLogTerm(v.termHistory) >= LastLogTerm(v2.termHistory)
+                              /\ (LastLogTerm(v.termHistory) = LastLogTerm(v2.termHistory) => v.flushLsn >= v2.flushLsn)
+       max_vote == prop_state[p].votes[max_vote_acc]
+       prop_th == Append(max_vote.termHistory, [term |-> prop_state[p].term, lsn |-> max_vote.flushLsn])
```

Comment:
We're copying up to flushLsn from the `max_vote_acc`, then declare us leader.
Where do we copy the wal to the other acceptors? (`--sync-safekeepers` in the implementation, I believe)

Maybe a comment would be appropriate here.

Reply:
I remember answering to this, but comment seems to be lost; long comment a couple of lines below explains this, expanded it a bit more.